File: public/api/admin/expand-tags.php
<?php
/**
 * 批量扩充本地标签库
 * - 拉 Danbooru 全量标签（按 post_count 排序）
 * - 翻译成中文（字典 → DB 已有 → MyMemory）
 * - 预下载每个标签的 top 1 示例图到本地
 *
 * 用法：
 *   POST /api/admin/expand-tags.php
 *   { "action": "start", "min_posts": 100, "max_pages": 50, "with_images": true }
 *   { "action": "status" }
 *   { "action": "stop" }
 *
 * 状态文件：storage/cache/expand-status.json（实时进度）
 * 索引：把已处理的 tag 名存到 storage/cache/expand-done.txt
 */
declare(strict_types=1);
require_once __DIR__ . '/../../../src/bootstrap.php';

use NaiStudio\Db;
use NaiStudio\TagDict;
use NaiStudio\Translator;

header('Content-Type: application/json; charset=utf-8');
header('Cache-Control: no-store');

$action = $_POST['action'] ?? $_GET['action'] ?? 'status';

// 状态文件位置
$stateFile = __DIR__ . '/../../../storage/cache/expand-status.json';
$doneFile  = __DIR__ . '/../../../storage/cache/expand-done.txt';
$previewDir = __DIR__ . '/../../../storage/tag-previews';
@mkdir(dirname($stateFile), 0777, true);
@mkdir($previewDir, 0777, true);

function readState(string $file): array {
    if (!file_exists($file)) return ['status' => 'idle', 'progress' => 0, 'total' => 0, 'message' => ''];
    $j = json_decode(file_get_contents($file), true);
    return is_array($j) ? $j : ['status' => 'idle', 'progress' => 0, 'total' => 0, 'message' => ''];
}

function writeState(string $file, array $state): void {
    file_put_contents($file, json_encode($state, JSON_UNESCAPED_UNICODE));
}

if ($action === 'status') {
    ok_response(readState($stateFile));
}

if ($action === 'stop') {
    $state = readState($stateFile);
    $state['status'] = 'stopped';
    $state['message'] = '已请求停止';
    writeState($stateFile, $state);
    ok_response($state);
}

if ($action === 'start') {
    // 异步执行（用 ignore_user_abort + fastcgi_finish_request 不行，普通 HTTP 同步）
    $minPosts = max(1, (int)($_POST['min_posts'] ?? 100));
    $maxPages = max(1, min(200, (int)($_POST['max_pages'] ?? 50)));  // 50 页 × 1000 = 50000
    $withImages = (bool)($_POST['with_images'] ?? true);

    // 设置超时（CLI 不受限制；web 请求给 300s）
    set_time_limit(0);

    // 已处理列表
    $done = [];
    if (file_exists($doneFile)) {
        $done = array_filter(array_map('trim', file($doneFile) ?: []));
        $done = array_flip($done);  // O(1) 查
    }

    // DB 中已有
    $existing = Db::fetchAll("SELECT name FROM tags", [], \PDO::FETCH_COLUMN);

    // 分类映射
    $catNames = ['General', 'Artist', 'Copyright', 'Character', 'Meta'];  // index 0/1/3/4/5
    $catIds = []; // Danbooru category → DB category_id（这里只关心 Danbooru 0=通用, 1=画师, 3=版权, 4=角色, 5=元数据）
    foreach (Db::fetchAll("SELECT id, name, name_cn FROM tag_categories") as $c) {
        $catIds[strtolower($c['name'])] = (int)$c['id'];
        if (!empty($c['name_cn'])) $catIds[strtolower($c['name_cn'])] = (int)$c['id'];
    }

    $state = [
        'status' => 'running',
        'started_at' => date('Y-m-d H:i:s'),
        'progress' => 0,
        'total' => $maxPages * 1000,
        'added' => 0,
        'translated' => 0,
        'images' => 0,
        'skipped' => 0,
        'errors' => 0,
        'current_tag' => '',
        'message' => '拉取 Danbooru 标签中...',
    ];
    writeState($stateFile, $state);

    // 关闭前端连接，让脚本继续
    if (function_exists('fastcgi_finish_request')) {
        // 输出 + 关闭
        echo json_encode(['ok' => true, 'message' => '已启动后台扩充'], JSON_UNESCAPED_UNICODE);
        fastcgi_finish_request();
    } else {
        echo json_encode(['ok' => true, 'message' => '已启动扩充', 'note' => '请勿关闭浏览器'], JSON_UNESCAPED_UNICODE);
    }

    // 拉取 + 处理
    $page = 1;
    $processed = 0;
    $state['total'] = $maxPages * 1000;
    $logFile = __DIR__ . '/../../../storage/cache/expand.log';
    @file_put_contents($logFile, "[" . date('H:i:s') . "] start min_posts={$minPosts} max_pages={$maxPages}\n", FILE_APPEND);

    while ($page <= $maxPages) {
        // 1) 拉这一页的标签（用 curl，PHP stream 在 Windows 上经常 timeout）
        //    Danbooru: search[order] 接受 date/count/name/similarity，按 post 排要 count
        $url = "https://danbooru.donmai.us/tags.json?limit=1000&page={$page}&search[order]=count";
        $body = httpGet($url, 30);
        if ($body === false) {
            $state['errors']++;
            $state['message'] = "第 {$page} 页拉取失败（网络问题）";
            @file_put_contents($logFile, "[" . date('H:i:s') . "] page {$page} fail\n", FILE_APPEND);
            writeState($stateFile, $state);
            $page++;
            usleep(2000000);
            continue;
        }
        @file_put_contents($logFile, "[" . date('H:i:s') . "] page {$page} got " . strlen($body) . " bytes\n", FILE_APPEND);
        // 调试：记第一条
        $firstTag = json_decode($body, true)[0] ?? null;
        if ($firstTag) @file_put_contents($logFile, "[" . date('H:i:s') . "] page {$page} first: " . $firstTag['name'] . " (" . ($firstTag['post_count'] ?? 0) . " posts)\n", FILE_APPEND);
        $tags = json_decode($body, true);
        if (!is_array($tags) || empty($tags)) {
            $state['message'] = "第 {$page} 页无数据，结束";
            break;
        }
        // 按 post_count 降序（Danbooru 的 search[order] 不稳）
        usort($tags, fn($a, $b) => ($b['post_count'] ?? 0) <=> ($a['post_count'] ?? 0));

        foreach ($tags as $t) {
            // 检查是否停止
            $curState = readState($stateFile);
            if ($curState['status'] === 'stopped') {
                $state['status'] = 'stopped';
                $state['message'] = "已停止（已处理 {$processed} 条）";
                writeState($stateFile, $state);
                exit(0);
            }

            $name = (string)($t['name'] ?? '');
            if ($name === '') continue;
            $postCount = (int)($t['post_count'] ?? 0);
            $cat = (int)($t['category'] ?? 0);
            $nsfw = ($t['is_deprecated'] ?? false) ? 1 : 0;

            if ($postCount < $minPosts) {
                $state['skipped']++;
                continue;
            }

            // 跳过已处理
            if (isset($done[$name])) {
                $state['skipped']++;
                continue;
            }

            $state['current_tag'] = $name;
            $state['progress'] = $processed;
            $processed++;

            // 2) 翻译
            $cn = TagDict::lookup($name);
            $translatedThisTag = false;
            if ($cn === null) {
                // 查 DB
                $existingCn = Db::fetchOne("SELECT cn_name FROM tags WHERE name = ?", [$name])['cn_name'] ?? null;
                if ($existingCn) {
                    $cn = $existingCn;
                } else {
                    // 调 MyMemory（限速 100ms 间隔）
                    usleep(100000);
                    $r = Translator::enToZh($name);
                    $cn = $r['cn'] ?? '';
                    if ($cn === '' || stripos($cn, 'WARNING') !== false) $cn = null;
                    else $translatedThisTag = true;
                }
            }

            // 3) 分类映射
            $catName = $catNames[$cat] ?? 'General';
            $catId = $catIds[strtolower($catName)] ?? null;
            if ($catId === null) {
                // fallback：找第一个
                $catId = (int)Db::fetchOne("SELECT id FROM tag_categories LIMIT 1")['id'];
            }

            // 4) 插入 DB
            try {
                $existing = Db::fetchOne("SELECT id FROM tags WHERE name = ?", [$name]);
                if ($existing) {
                    Db::update('tags', $existing['id'], [
                        'cn_name' => $cn,
                        'post_count' => $postCount,
                        'category_id' => $catId,
                        'is_nsfw' => $nsfw,
                    ]);
                } else {
                    Db::insert('tags', [
                        'name' => $name,
                        'category_id' => $catId,
                        'cn_name' => $cn,
                        'post_count' => $postCount,
                        'is_nsfw' => $nsfw,
                    ]);
                    $state['added']++;
                }
                if ($translatedThisTag) $state['translated']++;
            } catch (\Throwable $e) {
                $state['errors']++;
            }

            // 5) 预下载示例图
            if ($withImages && $postCount > 0) {
                $imgUrl = downloadExampleImage($name, $previewDir, $state);
                if ($imgUrl) {
                    try {
                        $row = Db::fetchOne("SELECT id FROM tags WHERE name = ?", [$name]);
                        if ($row) {
                            Db::update('tags', $row['id'], ['example_image_url' => $imgUrl]);
                            $state['images']++;
                        }
                    } catch (\Throwable $e) {}
                }
            }

            // 6) 记入 done
            file_put_contents($doneFile, $name . "\n", FILE_APPEND);
            $done[$name] = true;

            // 7) 进度（每 50 条写一次）
            if ($processed % 50 === 0) {
                $state['message'] = "处理中：第 {$page} 页，{$state['added']} 新增，{$state['translated']} 翻译，{$state['images']} 图";
                writeState($stateFile, $state);
            }

            // 限速
            usleep(20000); // 50/s
        }

        $page++;
        $state['message'] = "完成第 " . ($page - 1) . " 页，开始第 {$page} 页";
        writeState($stateFile, $state);
        usleep(200000); // 每页间隔
    }

    $state['status'] = 'done';
    $state['message'] = "完成！新增 {$state['added']} 条，翻译 {$state['translated']} 条，下载 {$state['images']} 张图";
    writeState($stateFile, $state);
    exit(0);
}

/**
 * 下载某个 tag 的 1 张示例图
 * @return string|null 相对 URL（存到 DB 用）
 */
function downloadExampleImage(string $tag, string $previewDir, array &$state): ?string {
    // 跳过：已有图
    $hash = substr(md5($tag), 0, 2);
    $subdir = "$previewDir/$hash";
    @mkdir($subdir, 0777, true);
    $fname = "$subdir/" . preg_replace('/[^a-z0-9_]/i', '_', $tag) . '.jpg';
    $relUrl = "/storage/tag-previews/$hash/" . basename($fname);

    if (file_exists($fname) && filesize($fname) > 1000) {
        return $relUrl;
    }

    // 拉 1 个 post（curl）
    $url = "https://danbooru.donmai.us/posts.json?tags=" . urlencode($tag) . "&limit=1&random=true";
    $body = httpGet($url, 15);
    if ($body === false) return null;
    $posts = json_decode($body, true);
    if (!is_array($posts) || empty($posts)) return null;

    $previewUrl = $posts[0]['preview_file_url'] ?? null;
    if (!$previewUrl) return null;

    // 下载（curl）
    $img = httpGet($previewUrl, 20);
    if (strlen($img) < 500) return null;
    file_put_contents($fname, $img);
    return $relUrl;
}

/**
 * 用 curl 拉取 URL（比 file_get_contents 在 Windows 上可靠）
 */
function httpGet(string $url, int $timeout = 30): string {
    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_TIMEOUT        => $timeout,
        CURLOPT_CONNECTTIMEOUT => min(10, $timeout),
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_USERAGENT      => 'NAI-Studio/1.0 (local)',
        CURLOPT_SSL_VERIFYPEER => false,
    ]);
    $body = curl_exec($ch);
    if (curl_errno($ch)) {
        error_log('[expand-tags] httpGet error: ' . curl_error($ch) . ' url=' . $url);
    }
    curl_close($ch);
    return $body === false ? '' : (string)$body;
}

File: src-tauri/src/api/admin/expand_tags.rs
//! /api/admin/expand-tags -- 批量扩充本地标签库
//!
//! 跟 NAI Studio PHP 项目 admin/expand-tags.php 等价
//!   GET  /api/admin/expand-tags        -> 状态
//!   POST /api/admin/expand-tags        -> 启动 (body: {min_posts, max_pages, with_images})
//!   DELETE /api/admin/expand-tags      -> 停止
//!
//! 行为:
//!   - 拉 Danbooru tags (按 post_count 排序,默认 min_posts=100 过滤低频)
//!   - 翻译链: 内置字典 (秒回) -> DB 已有 cn_name -> MyMemory 公开 API (限速)
//!   - 预下载每个 tag 的 top 1 示例图到 storage/tag-previews/{hash}/

use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Duration;

use axum::Json;
use axum::extract::State;
use chrono::Utc;
use reqwest::Client;
use rusqlite::params;
use serde_json::{Value, json};
use tokio::time::sleep;

use super::{JobState, get_or_create, is_cancelled, update};
use crate::api::SharedState;
use crate::error::AppResult;

const NAME: &str = "expand_tags";
const DANBOORU_BASE: &str = "https://danbooru.donmai.us";
const MYMEMORY_URL: &str = "https://api.mymemory.translated.net/get";

pub async fn status(State(_state): State<SharedState>) -> AppResult<Json<Value>> {
    let j = get_or_create(NAME);
    Ok(Json(j.to_json()))
}

pub async fn start(
    State(state): State<SharedState>,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let j = get_or_create(NAME);
    if j.status == "running" {
        return Ok(Json(json!({"ok": false, "error": "已经在运行"})));
    }
    let min_posts: i64 = body.get("min_posts").and_then(|v| v.as_i64()).unwrap_or(100).max(1);
    let max_pages: i64 = body.get("max_pages").and_then(|v| v.as_i64()).unwrap_or(20).clamp(1, 200);
    let with_images: bool = body.get("with_images").and_then(|v| v.as_bool()).unwrap_or(true);

    let _ = j.cancel.store(false, Ordering::Relaxed);
    update(NAME, |j| {
        j.status = "running".to_string();
        j.started_at = Some(std::time::Instant::now());
        j.finished_at = None;
        j.done = 0;
        j.total = max_pages * 1000;
        j.message = format!("启动：min_posts={} max_pages={} with_images={}", min_posts, max_pages, with_images);
    });
    let cancel = j.cancel.clone();
    tokio::spawn(async move {
        run_worker(state, min_posts, max_pages, with_images, cancel).await;
    });
    Ok(Json(json!({"ok": true, "message": "已启动"})))
}

pub async fn stop(State(_state): State<SharedState>) -> AppResult<Json<Value>> {
    let j = get_or_create(NAME);
    j.cancel.store(true, Ordering::Relaxed);
    update(NAME, |j| {
        j.status = "stopped".to_string();
        j.message = "已请求停止".to_string();
        j.finished_at = Some(std::time::Instant::now());
    });
    Ok(Json(json!({"ok": true, "message": "已停止"})))
}

async fn run_worker(
    state: SharedState,
    min_posts: i64,
    max_pages: i64,
    with_images: bool,
    cancel: Arc<AtomicBool>,
) {
    // 1. 已处理列表 (跨进程跳过) — 写到 cache 文件
    let done_path = state.paths.cache.join("expand-done.txt");
    let _ = std::fs::create_dir_all(&state.paths.cache);
    let mut done: HashMap<String, ()> = HashMap::new();
    if let Ok(s) = std::fs::read_to_string(&done_path) {
        for line in s.lines() {
            done.insert(line.trim().to_string(), ());
        }
    }
    let mut done_file = std::fs::OpenOptions::new()
        .create(true).append(true).open(&done_path).ok();

    // 2. 分类映射 (Danbooru category index 0/1/3/4/5 -> DB tag_categories.id)
    let cat_map: HashMap<String, i64> = {
        let conn = state.db.lock();
        let mut stmt = match conn.prepare("SELECT slug, id FROM tag_categories") {
            Ok(s) => s,
            Err(e) => {
                update(NAME, |j| { j.status = "error".into(); j.message = format!("DB: {}", e); j.finished_at = Some(std::time::Instant::now()); });
                return;
            }
        };
        let rows = stmt.query_map([], |r| Ok((r.get::<_, String>(0)?, r.get::<_, i64>(1)?))).ok();
        let map: HashMap<String, i64> = rows.map(|rs| rs.filter_map(|r| r.ok()).collect()).unwrap_or_default();
        map
    };
    let default_cat = cat_map.get("general").copied().unwrap_or(1);

    // 3. HTTP client (含代理)
    let client = match state.proxy_url() {
        Some(p) => {
            let proxy = reqwest::Proxy::all(&p).ok();
            match proxy {
                Some(pr) => reqwest::Client::builder()
                    .timeout(Duration::from_secs(30))
                    .user_agent("nai-studio-desktop/2.0 (expand-tags)")
                    .proxy(pr)
                    .build(),
                None => reqwest::Client::builder()
                    .timeout(Duration::from_secs(30))
                    .user_agent("nai-studio-desktop/2.0 (expand-tags)")
                    .build(),
            }
        }
        None => reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .user_agent("nai-studio-desktop/2.0 (expand-tags)")
            .build(),
    };
    let client = match client {
        Ok(c) => c,
        Err(e) => {
            update(NAME, |j| { j.status = "error".into(); j.message = format!("reqwest: {}", e); j.finished_at = Some(std::time::Instant::now()); });
            return;
        }
    };

    let mut page = 1i64;

    'pages: while page <= max_pages {
        if cancel.load(Ordering::Relaxed) || is_cancelled(NAME) { break; }

        update(NAME, |j| { j.message = format!("拉第 {}/{} 页...", page, max_pages); j.current_page = page; j.pages_total = max_pages; });

        // 拉这一页
        let url = format!(
            "{}/tags.json?limit=1000&page={}&search[order]=count",
            DANBOORU_BASE, page
        );
        let resp = match client.get(&url).send().await {
            Ok(r) => r,
            Err(e) => {
                update(NAME, |j| { j.errors += 1; j.last_error = format!("page {} net: {}", page, e); j.message = format!("第 {} 页网络错误: {}", page, e); });
                sleep(Duration::from_secs(2)).await;
                page += 1;
                continue;
            }
        };
        if !resp.status().is_success() {
            update(NAME, |j| { j.errors += 1; j.last_error = format!("page {} HTTP {}", page, resp.status().as_u16()); j.message = format!("第 {} 页 HTTP {}", page, resp.status().as_u16()); });
            sleep(Duration::from_secs(2)).await;
            page += 1;
            continue;
        }
        let tags: Vec<Value> = resp.json().await.unwrap_or_default();
        if tags.is_empty() {
            update(NAME, |j| { j.message = format!("第 {} 页空数据, 结束", page); });
            break;
        }

        for t in &tags {
            if cancel.load(Ordering::Relaxed) || is_cancelled(NAME) { break 'pages; }

            let name = t.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
            if name.is_empty() || name.len() > 128 { update(NAME, |j| { j.skipped += 1; }); continue; }
            let post_count = t.get("post_count").and_then(|v| v.as_i64()).unwrap_or(0);
            if post_count < min_posts { update(NAME, |j| { j.skipped += 1; }); continue; }
            if done.contains_key(&name) { update(NAME, |j| { j.skipped += 1; }); continue; }

            update(NAME, |j| {
                j.done += 1;
                j.current_tag = name.clone();
                j.message = format!("处理 {} ({} post)", name, post_count);
            });

            // 翻译: 内置字典 -> DB 已有 -> MyMemory
            let cn_name: Option<String> = if let Some(cn) = super::super::decompose::builtin_dict_pub(&name) {
                Some(cn.to_string())
            } else {
                let from_db = {
                    let conn = state.db.lock();
                    conn.query_row(
                        "SELECT cn_name FROM tags WHERE name = ?1",
                        params![name],
                        |r| r.get::<_, Option<String>>(0)
                    ).ok().flatten()
                };
                match from_db {
                    Some(cn) if !cn.is_empty() => Some(cn),
                    _ => {
                        sleep(Duration::from_millis(120)).await;  // 限速
                        if let Some(cn) = translate_en_to_zh(&client, &name).await {
                            update(NAME, |j| { j.translated += 1; });
                            Some(cn)
                        } else {
                            None
                        }
                    }
                }
            };

            // 分类映射
            let cat_idx = t.get("category").and_then(|v| v.as_i64()).unwrap_or(0);
            let cat_slug = match cat_idx {
                1 => "artist",
                3 => "copyright",
                4 => "character",
                5 => "meta",
                _ => "general",
            };
            let cat_id = cat_map.get(cat_slug).copied().unwrap_or(default_cat);

            let is_nsfw = if t.get("is_deprecated").and_then(|v| v.as_bool()).unwrap_or(false) { 1 } else { 0 };

            // 入库
            let now = Utc::now().to_rfc3339();
            let result = {
                let conn = state.db.lock();
                conn.execute(
                    "INSERT INTO tags (name, category_id, cn_name, post_count, is_nsfw, created_at)
                     VALUES (?1, ?2, ?3, ?4, ?5, ?6)
                     ON CONFLICT(name) DO UPDATE SET
                        category_id = excluded.category_id,
                        cn_name = COALESCE(NULLIF(tags.cn_name, ''), excluded.cn_name),
                        post_count = excluded.post_count,
                        is_nsfw = excluded.is_nsfw",
                    params![name, cat_id, cn_name, post_count, is_nsfw, now],
                )
            };
            if result.is_err() {
                update(NAME, |j| { j.errors += 1; });
            } else {
                update(NAME, |j| { j.added += 1; });
                if let Some(ref mut f) = done_file {
                    use std::io::Write;
                    let _ = writeln!(f, "{}", name);
                }
                done.insert(name.clone(), ());
            }

            // 预下载示例图
            if with_images && post_count > 0 {
                if let Some(rel_url) = download_example_image(&state, &client, &name).await {
                    let conn = state.db.lock();
                    let _ = conn.execute(
                        "UPDATE tags SET example_image_url = ?1 WHERE name = ?2",
                        params![rel_url, name],
                    );
                    update(NAME, |j| { j.images += 1; });
                }
                sleep(Duration::from_millis(60)).await;  // 限速
            }
        }
        page += 1;
        sleep(Duration::from_millis(200)).await;  // 页间间隔
    }

    let final_status = if cancel.load(Ordering::Relaxed) || is_cancelled(NAME) { "stopped" } else { "done" };
    update(NAME, |j| {
        j.status = final_status.to_string();
        j.message = format!("完成: 新增 {} 翻译 {} 图 {} 跳过 {} 错 {}",
            j.added, j.translated, j.images, j.skipped, j.errors);
        j.finished_at = Some(std::time::Instant::now());
    });
    let _ = cancel;
}

/// MyMemory en -> zh-CN
async fn translate_en_to_zh(client: &Client, name: &str) -> Option<String> {
    let url = format!("{}?q={}&langpair=en|zh-CN", MYMEMORY_URL, urlencoded(name));
    let resp = client.get(&url).send().await.ok()?;
    if !resp.status().is_success() { return None; }
    let data: Value = resp.json().await.ok()?;
    let translated = data.get("responseData")
        .and_then(|rd| rd.get("translatedText"))
        .and_then(|t| t.as_str())
        .map(|s| s.trim().to_string())?;
    if translated.is_empty() { return None; }
    if translated.to_lowercase() == name.to_lowercase() { return None; }
    // MyMemory 失败时常返 "PLEASE SELECT TWO DISTINCT LANGUAGES" 等
    if translated.contains("PLEASE SELECT") || translated.contains("WARNING") { return None; }
    Some(translated.chars().take(128).collect())
}

async fn download_example_image(
    state: &SharedState,
    client: &Client,
    tag: &str,
) -> Option<String> {
    let hash = {
        let mut h: u32 = 5381;
        for b in tag.bytes() { h = h.wrapping_mul(33).wrapping_add(b as u32); }
        format!("{:02x}", h & 0xFF)
    };
    let safe_name: String = tag.chars()
        .map(|c| if c.is_ascii_alphanumeric() || c == '_' { c } else { '_' })
        .collect();
    let subdir: PathBuf = state.paths.tag_previews.join(&hash);
    let _ = std::fs::create_dir_all(&subdir);
    let fname = subdir.join(format!("{}.jpg", safe_name));
    let rel_url = format!("/storage/tag-previews/{}/{}.jpg", hash, safe_name);

    // 本地已有
    if let Ok(meta) = std::fs::metadata(&fname) {
        if meta.len() > 1000 { return Some(rel_url); }
    }

    // 拉 1 个随机 post
    let url = format!("{}/posts.json?tags={}&limit=1&random=true", DANBOORU_BASE, urlencoded(tag));
    let resp = client.get(&url).send().await.ok()?;
    if !resp.status().is_success() { return None; }
    let posts: Vec<Value> = resp.json().await.ok()?;
    let preview_url = posts.first()
        .and_then(|p| p.get("preview_file_url"))
        .and_then(|v| v.as_str())?;
    if preview_url.is_empty() { return None; }

    // 下载
    let img = client.get(preview_url).send().await.ok()?;
    if !img.status().is_success() { return None; }
    let bytes = img.bytes().await.ok()?;
    if bytes.len() < 500 { return None; }
    if std::fs::write(&fname, &bytes).is_err() { return None; }
    Some(rel_url)
}

fn urlencoded(s: &str) -> String {
    s.chars().map(|c| {
        if c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.' || c == '~' {
            c.to_string()
        } else {
            format!("%{:02X}", c as u8)
        }
    }).collect()
}

#[allow(dead_code)]
fn _phantom(_: JobState) {}
